File: algoritmito.py
from icecream import ic
from enum import Enum, auto

"""
	logica de recuperatorio: se toman 4 temas y si o si tienen que estar bien los 4 temas
"""


"""
	
"""


"""
igual o mayor a 4 final. no hay promocion.


si la nota del promedio de los dos parciales es mayor 8:
	parcial = 1 tema.

-si la nota del promedio 
ajhkajakjhfadiufhakjfh

dos parciales
	un parcial tiene 8 temas
		cada tema 4 notas (a,b,c,d)

final 5 temas
alumnos 100
"""
	
# //-----------------------------------------------------------------------//
# ;;----------------------ok. classes.--------------------------;;
# //-----------------------------------------------------------------------//
	
class NotaEnLetra(Enum):
	A = 9 
	B = 7
	C = 5
	D = 3
	E = 1

class TemaDeParcial:
	def __init__(self, name, nota):
		self.name = name
		self.nota = nota
		self.aprobado = nota.value > NotaEnLetra.D.value
		self.desaprobado = not self.aprobado
	
	def __repr__(self):
		return f"|{self.name}|Nota:{self.nota}|"
		
	def __lt__(self, other):
		return self.nota.value < other.nota.value

	
class Parcial:
	def __init__(self, temario, temas_notas):
		self.temario = temario
		self.temas_notas = temas_notas
		self.temas_aprobados = [temadeparcial for temadeparcial in self.temas_notas if temadeparcial.aprobado]
		self.temas_desaprobados = [temadeparcial for temadeparcial in self.temas_notas if temadeparcial.desaprobado]
		self.nota = (len(self.temas_aprobados) / len(self.temas_notas)) * (10)

	def __str__(self):
		return f"Parcial de {self.temario}\n\tTemas aprobados: {len(self.temas_aprobados)}\n\tTemas desaprobados: {len(self.temas_desaprobados)}\n\tContenidos desaprobados: {self.temas_desaprobados}\n\tNota: {self.nota} "
class Alumno:
	_temas_de_final = None
	def __init__(self, name, parcial1, parcial2):
		self.name = name
		self.parcial1 = parcial1
		self.parcial2 = parcial2
		self.promedio = (self.parcial1.nota + self.parcial2.nota)/2
		self.recursa = self.parcial1.nota < 4 or self.parcial2.nota < 4 
		self.temas_aprobados = sorted(self.parcial1.temas_aprobados + self.parcial2.temas_aprobados)
		self.temas_desaprobados = sorted(self.parcial1.temas_desaprobados + self.parcial2.temas_desaprobados)
		# ic(
			# [ tema for tema in self.temas_aprobados if tema.nota.value < NotaEnLetra.B.value ]
		# )
		# self.promociona = len(self.temas_desaprobados) == 0 and not any([tema for tema in self.temas_aprobados if tema.nota.value < NotaEnLetra.B.value])
		self.promociona = not bool([ tema for tema in self.temas_aprobados if tema.nota.value < NotaEnLetra.B.value ])
		# self.promociona = False
		
		
	@property
	def temas_de_final(self):
		if self._temas_de_final is None:
			if self.promociona:
				self._temas_de_final = "Promociona"
			elif self.recursa:
				self._temas_de_final = "Recursa"
			elif self.temas_desaprobados and self.promedio >= 8:
				self._temas_de_final = self.temas_desaprobados[0]
			else:
				self._temas_de_final = self.temas_desaprobados
				
			if len(self.temas_desaprobados) > 5:
				self._temas_de_final = self.temas_desaprobados[0:5]
			
		return self._temas_de_final
		
	
	def __str__(self):
		return f"Alumno {self.name}:\n   Parcial 1: {self.parcial1}\n   Parcial 2: {self.parcial2}\n   Promedio: {self.promedio}\n   Temas para el final: {self.temas_de_final}"
	








# //--------------------------------------------------------------------//
# ;;---------------------------ok. iniciar-----------------------;;
# //--------------------------------------------------------------------//

if __name__ == "__main__":
	parcial1 = Parcial(
			temario="Algoritmos 1", 
			temas_notas = [
				TemaDeParcial("tema1", NotaEnLetra.D),
				TemaDeParcial("tema2", NotaEnLetra.A),
				TemaDeParcial("tema3", NotaEnLetra.A),
				TemaDeParcial("tema4", NotaEnLetra.A),
				TemaDeParcial("tema5", NotaEnLetra.A),
				TemaDeParcial("tema6", NotaEnLetra.A),
				TemaDeParcial("tema7", NotaEnLetra.A),
				TemaDeParcial("tema8", NotaEnLetra.A),
			]
	)
	parcial2 = Parcial(
			temario="Algoritmos 2", 
			temas_notas={
				TemaDeParcial("tema1", NotaEnLetra.D),
				TemaDeParcial("tema2", NotaEnLetra.D),
				TemaDeParcial("tema3", NotaEnLetra.D),
				TemaDeParcial("tema4", NotaEnLetra.B),
				TemaDeParcial("tema5", NotaEnLetra.A),
				TemaDeParcial("tema6", NotaEnLetra.B),
				TemaDeParcial("tema7", NotaEnLetra.A),
				TemaDeParcial("tema8", NotaEnLetra.C),
			}
	)
	alex = Alumno("Ayelen", parcial1, parcial2)
	print(alex)
	ic(alex.promociona)

File: algoritmito.cpp
#include <iostream>
#include <string>
using namespace std;

//--------------------------------------------------------------------//
// ;;---------------------------ok. enums-----------------------;;
//--------------------------------------------------------------------//

enum NotaEnLetra {
    A = 9,
    B = 7,
    C = 5,
    D = 3,
    E = 1
};

//--------------------------------------------------------------------//
// ;;---------------------------ok. classes-----------------------;;
//--------------------------------------------------------------------//

//-------------------------ok.Parcial-----------------------//
class TemaDeParcial {
  public:
    string name;
    NotaEnLetra nota;
    bool aprobado;
    bool desaprobado;
	TemaDeParcial ( string name, NotaEnLetra nota){
		this->name = name;
		this->nota = nota;
		aprobado = this->nota > NotaEnLetra::D;
		desaprobado = !aprobado;
	};
};

//-------------------------ok.Parcial-----------------------//
class Parcial {
  public:
	string temario;
    // TemaDeParcial* temas_notas[8];
	TemaDeParcial* temas_aprobados;
	TemaDeParcial* temas_desaprobados;
	int nota;
	int n_desaprobados;
	int n_aprobados;
	
	Parcial(string temario, TemaDeParcial temas_notas[8]){
		this->temario = temario;
		// this->temas_notas = temas_notas;	// just save locally the direction, no need copy
        // for (int i = 0; i < 8; i++) {
            // this->temas_notas[i] = &temas_notas[i];
        // }
		
        // Separar aprobados y desaprobados counts
		n_desaprobados = 0;
		for ( int i = 0; i < 8; i++ ) {
			if (temas_notas[i].desaprobado) {
				n_desaprobados++;
			}
		}
		n_aprobados = 8 - n_desaprobados;

		// Allocate arrays based on counts
		temas_desaprobados = new TemaDeParcial[n_desaprobados];
		temas_aprobados = new TemaDeParcial[n_aprobados];


        // Llenar los arrays respectivamente		
        int desaprobados_index = 0;
        int aprobados_index = 0;
        for (int i = 0; i < 8; i++) {
            if (temas_notas[i].desaprobado) {
                temas_desaprobados[desaprobados_index++] = temas_notas[i];
            } else {
                temas_aprobados[aprobados_index++] = temas_notas[i];
            }
        }
		
		// figure out nota
		// nota = (n_aprobados / n_desaprobados) * 10;
		nota = (n_aprobados > 0) ? (10 * n_aprobados / 8) : 0;
	}

	~Parcial() {
		delete[] temas_desaprobados;
		delete[] temas_aprobados;
	}
};

//-------------------------ok.Parcial-----------------------//
class Alumno {
	string name;
	Parcial* parcial1;
	Parcial* parcial2;
	float promedio;
	bool recursa;
	bool promociona;
	int n_desaprobados;
	int n_aprobados;
	TemaDeParcial* temas_desaprobados;
	TemaDeParcial* temas_aprobados;
	TemaDeParcial** temas_de_final; // Array of pointers to TemaDeParcial
	
  public:
	Alumno( string name, Parcial* parcial1, Parcial* parcial2 )
		: (name(name), parcial1(parcial1), parcial2(parcial2)) {
			
		// sacar el promedio de parciales
		promedio = (parcial1->nota + parcial2->nota) / 2.0;
		recursa = (parcial1->nota < 4) || (parcial2->nota < 4);
		
		// separar aprobados y desaprobados counts
		n_desaprobados = parcial1->n_desaprobados + parcial2->n_desaprobados;
		n_aprobados = parcial1->n_aprobados + parcial2->n_aprobados;
		
		// Allocate arrays based on counts
		temas_desaprobados = new TemaDeParcial[n_desaprobados];
		temas_aprobados = new TemaDeParcial[n_aprobados];

		// Llenar my temas_desaprobados array
		for (int i = 0; ; i < parcial1->n_desaprobados; i++ ) {
			this->temas_desaprobados[i] = parcial1->temas_desaprobados[i]
		}
		for (int i = 0; ; i < parcial2->n_desaprobados+i; i++ ) {
			this->temas_desaprobados[i+parcial1->n_desaprobados] = parcial2->temas_desaprobados[i]
		}
		
		// Llenar my temas_aprobados array
		for (int i = 0; ; i < parcial1->n_desaprobados; i++ ) {
			this->temas_aprobados[i] = parcial1->temas_aprobados[i]
		}
		for (int i = 0; ; i < parcial2->n_desaprobados+i; i++ ) {
			this->temas_aprobados[i+parcial1->n_desaprobados] = parcial2->temas_aprobados[i]
		}
		
		// averiguar si promociona
		promociona = true;
		for ( int i = 0; i < n_aprobados; i++ ) {
			if (temas_aprobados[i].nota < NotaEnLetra::B){
				promociona = false;
				break;
			}
		}
		
		
		
		// averiguar la cantidad de temas para el final
		if (promociona || recursa) {
			n_temas_de_final = 0;
		}
		else if (n_desaprobados && promedio >= 8){
			n_temas_de_final = 1;
		} 
		else if (n_desaprobados > 5) {
			n_temas_de_final = 5;
		}
		else {
			n_temas_de_final = n_desaprobados;
		}
		

        // Allocate and fill temas_de_final
        temas_de_final = new TemaDeParcial*[n_temas_de_final];
        if (n_temas_de_final > 0) {
            if (n_desaprobados > 0 && promedio >= 8) {
                temas_de_final[0] = &temas_desaprobados[0];
            } else {
                for (int i = 0; i < n_temas_de_final; i++) {
                    temas_de_final[i] = &temas_desaprobados[i];
                }
            }
        }
	}
	~Alumno() {
		// Clean up dynamically allocated memory
		delete[] temas_desaprobados;
		delete[] temas_aprobados;
		delete[] temas_de_final;
	}
	
	void print() {
		cout << "[Alumno " << name << "]" << endl;
		cout << "Promedio: " << promedio << endl;
		cout << "Recursa: " << (recursa ? "Si" : "No") << endl;
		cout << "Promociona: " << (promociona ? "Si" : "No") << endl;
		cout << "Numero de desaprobados: " << n_desaprobados << endl;
		cout << "Numero de aprobados: " << n_aprobados << endl;

		cout << "Temas desaprobados:" << endl;
		for (int i = 0; i < n_desaprobados; i++) {
			cout << temas_desaprobados[i].tema << " - Nota: " << temas_desaprobados[i].nota << endl;
		}

		cout << "Temas aprobados:" << endl;
		for (int i = 0; i < n_aprobados; i++) {
			cout << temas_aprobados[i].tema << " - Nota: " << temas_aprobados[i].nota << endl;
		}

		cout << "Temas para el final:" << endl;
		for (int i = 0; temas_de_final != nullptr && temas_de_final[i] != nullptr; i++) {
			cout << temas_de_final[i]->tema << " - Nota: " << temas_de_final[i]->nota << endl;
		}
	}
};





//--------------------------------------------------------------------//
// ;;---------------------------ok. iniciar-----------------------;;
//--------------------------------------------------------------------//

int main() {
	// primer parcial: 8 temas
	TemaDeParcial parcial1_temas[8] = {
		TemaDeParcial("tema1", NotaEnLetra::D),
		TemaDeParcial("tema2", NotaEnLetra::A),
		TemaDeParcial("tema3", NotaEnLetra::A),
		TemaDeParcial("tema4", NotaEnLetra::A),
		TemaDeParcial("tema5", NotaEnLetra::A),
		TemaDeParcial("tema6", NotaEnLetra::A),
		TemaDeParcial("tema7", NotaEnLetra::A),
		TemaDeParcial("tema8", NotaEnLetra::A)
	};
	Parcial parcial1("Estadistica 1", parcial1_temas);
	
	// segundo parcial: 8 temas
	TemaDeParcial parcial2_temas[8] = {
		TemaDeParcial("tema1", NotaEnLetra::D),
		TemaDeParcial("tema2", NotaEnLetra::A),
		TemaDeParcial("tema3", NotaEnLetra::A),
		TemaDeParcial("tema4", NotaEnLetra::A),
		TemaDeParcial("tema5", NotaEnLetra::A),
		TemaDeParcial("tema6", NotaEnLetra::A),
		TemaDeParcial("tema7", NotaEnLetra::A),
		TemaDeParcial("tema8", NotaEnLetra::A)
	};
	Parcial parcial2("Estadistica 2", parcial2_temas);
	
	
	// instace alumnito
	Alumno alumno1("Alex", &parcial1, &parcial2);
	
	
	// show his stuff thru his method
	alumno1.print();
	
	// Clean up dynamically allocated memory
	for (int i = 0; i < 8; i++) {
		delete parcial1_temas[i];
		delete parcial2_temas[i];
	}
	
	return 0;
}